### PR TITLE
Do not use f64 for STORAGE_REBATE_RATE

### PR DESCRIPTION
**Pull request by dariorussi**
as in https://github.com/MystenLabs/sui/pull/8102
the idea is to remove the f64 to be on the safer side

### CODE REVIEW

File: crates/sui-protocol-constants/src/lib.rs [hunk 1]
```diff
@@ -105,9 +105,12 @@ pub const OBJ_METADATA_COST_NON_REFUNDABLE: u64 = 50;
...
 /// === Tokenomics ===
 
-// TODO: this should be changed to u64.
 /// Sender of a txn that touches an object will get 99% of the storage rebate back.
-pub const STORAGE_REBATE_RATE: f64 = 0.99;
+///
+/// STORAGE_REBATE_RATE is a pair of integers, representing a fraction of a percentage
+/// value. Use for deterministic computation with a multiplication of the first value
+/// divided by the second. Values can have any number of digits and precision.
+pub const STORAGE_REBATE_RATE: (u64, u64) = (99, 100);
```
Comment (dariorussi) [inline crates/sui-protocol-constants/src/lib.rs:113]: ridiculously verbose comment....

File: crates/sui-types/src/temporary_store.rs [hunk 2]
```diff
@@ -670,7 +670,19 @@ impl<S> TemporaryStore<S> {
...
+        // multiply by STORAGE_REBATE_RATE first value (numerator)
+        let mut gas_rebate =
+            (cost_summary.storage_rebate as u128 * STORAGE_REBATE_RATE.0 as u128) as u64;
+        // divide by STORAGE_REBATE_RATE second value and round
+        gas_rebate = ((gas_rebate as u128 + (STORAGE_REBATE_RATE.1 as u128 / 2))
+            / STORAGE_REBATE_RATE.1 as u128) as u64;
+        // this is here in case we want to truncate
+        // gas_rebate = (gas_rebate as u128 / STORAGE_REBATE_RATE.1 as u128) as u64;
+        // REMOVE THIS (and related), here just to check if there are failures
+        assert_eq!(gas_rebate, gas_rebate1);
```
Comment (dariorussi) [inline crates/sui-types/src/temporary_store.rs:685]: this and related code will go away but it was to make sure we do not hit the assert which we did before I had rounding proper. I am not sure if you should truncate instead, pretty minor but wondering if people have opinions...

File: crates/sui-types/src/temporary_store.rs [hunk 3]
```diff
@@ -670,7 +670,19 @@ impl<S> TemporaryStore<S> {
...
         let gas_used = cost_summary.gas_used();
-        let gas_rebate = (cost_summary.storage_rebate as f64 * STORAGE_REBATE_RATE).round() as u64;
+        // value as computed before
+        let gas_rebate1 = (cost_summary.storage_rebate as f64 * 0.99).round() as u64;
+        // computation with u64 arithmetic
+        // multiply by STORAGE_REBATE_RATE first value (numerator)
+        let mut gas_rebate =
+            (cost_summary.storage_rebate as u128 * STORAGE_REBATE_RATE.0 as u128) as u64;
+        // divide by STORAGE_REBATE_RATE second value and round
+        gas_rebate = ((gas_rebate as u128 + (STORAGE_REBATE_RATE.1 as u128 / 2))
```
Comment (emmazzz) [inline crates/sui-types/src/temporary_store.rs:680]: Why not just either rounding up or rounding down? I don't have strong preference but this line feels a little confusing.

File: crates/sui-protocol-constants/src/lib.rs (hunk 1 above)
Comment (emmazzz) [reply crates/sui-protocol-constants/src/lib.rs:113]: Other rates in this file are in basis point. Do we want to be consistent here too?

File: crates/sui-types/src/temporary_store.rs (hunk 3 above)
Comment (dariorussi) [reply crates/sui-types/src/temporary_store.rs:680]: we can do both and break the previous behavior, which is fine by me but I would love for us to think about it.
Bottom line the old code has a `round()` call, some test rely on it and I am not sure we what we want to do.
You'll notice the next line (commented) is asking exactly that question

File: crates/sui-protocol-constants/src/lib.rs (hunk 1 above)
Comment (dariorussi) [reply crates/sui-protocol-constants/src/lib.rs:113]: I do not know and good point. It was not that, but that does not mean it should not be.
To me it's unclear what that value wants to be and what are the possible variation.

Comment (dariorussi) [reply crates/sui-protocol-constants/src/lib.rs:113]: I am actually liking the idea and I will move the code to do that.
After all when we are changing those values we are updating the code for now, so it does not really matter....

File: crates/sui-types/src/temporary_store.rs (hunk 3 above)
Comment (dariorussi) [reply crates/sui-types/src/temporary_store.rs:680]: I will add that my vote is for truncation, the simplest and most clear way to do it. I believe also more reliable as the value rounding down would less likely create issue with gas

Comment (dariorussi) [reply crates/sui-types/src/temporary_store.rs]: moved to truncation

Comment (dariorussi) [reply crates/sui-types/src/temporary_store.rs]: and back to rounding which makes things working and easy

**Comment by dariorussi** [comment]
@emmazzz take a look. I tried removing the rounding to use truncation but I got a bunch of errors in tests and it was a bit too much for me to change. It did not seem to be worth the effort.
I splashed a comment there that should be enough to clarify.
Please take a look

**Review by emmazzz — APPROVED** [review]

**Pull request merged**